### PR TITLE
Scope & Collections API - without implementation - ObjC 

### PR DESCRIPTION
**Pull request by jayahariv**
* Scope and Collection APIs without implementation ObjC. 
* Deprecate the previous APIs, as per the spec. 
* Unit tests are added to make sure, those APIs can be called and won't be crashing.

### CODE REVIEW

File: Objective-C/CBLQuery.mm [hunk 1]
```diff
@@ -79,6 +79,21 @@ - (instancetype) initWithDatabase: (CBLDatabase*)database
...
+- (instancetype) initWithCollection: (CBLCollection*)collection
+                 JSONRepresentation: (NSData*)json
+{
+    Assert(collection);
+    Assert(json);
+    self = [super init];
+    if (self) {
+        _json = json;
+        _language = kC4JSONQuery;
+
```
Comment (jayahariv) [inline Objective-C/CBLQuery.mm:91]:
```suggestion
```

File: Objective-C/CBLReplicatorConfiguration.h [hunk 2]
```diff
@@ -191,6 +215,30 @@ typedef BOOL (^CBLReplicationFilter) (CBLDocument* document, CBLDocumentFlags fl
...
 - (instancetype) initWithConfig: (CBLReplicatorConfiguration*)config;
 
+/**
+ Add a collection used for the replication with an optional collection configuration.
+ If the collection has been added before, the previous added and its configuration if specified
+ will be replaced.
+ If a null configuration is specified, a default empty configuration will be applied.
+ 
+ @param collection The collection to be added.
+ @param config Respective configuration for the collection, if nil, default config */
```
Comment (pasin) [inline Objective-C/CBLReplicatorConfiguration.h:225]: Maybe just Configuration for the collection?

File: Objective-C/CBLScope.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+//  limitations under the License.
+//
+
+#import <Foundation/Foundation.h>
+
+@class CBLCollection;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol CBLScope <NSObject>
```
Comment (pasin) [inline Objective-C/CBLScope.h:26]: Missing API doc.

File: Objective-C/CBLScope.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma mark Collections
+
+/** Get all collections in the scope. */
+- (NSArray<CBLCollection*>*) getCollections;
+
+/**
+ Get a collection in the scope by name.
+ If the collection doesn't exist, a nil value will be returned. */
+- (CBLCollection*) getCollectionWithName: (NSString*)name;
```
Comment (pasin) [inline Objective-C/CBLScope.h:41]: collectionWithName:

File: Objective-C/CBLScope.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma mark Properties
+
+/** Scope name. */
+@property (readonly, nonatomic) NSString* name;
+
+#pragma mark Collections
+
+/** Get all collections in the scope. */
+- (NSArray<CBLCollection*>*) getCollections;
```
Comment (pasin) [inline Objective-C/CBLScope.h:36]: collections

File: Objective-C/CBLScope.h (hunk 3 above)
Comment (pasin) [inline Objective-C/CBLScope.h:26]: I think it's unnecessary for iOS to have protocol for CBLScope and Could be CBLCollection as well. 

Even though the spec said "Protocol", it's up to the platform to decide whether to have it or not. I believe Java will not have it as protocol. Let evaluate this when you implement it.

**Review by pasin — CHANGES_REQUESTED** [review]

File: Objective-C/CBLQuery.mm [hunk 6]
```diff
@@ -182,7 +197,12 @@ - (instancetype) initWithSelect: (NSArray<CBLQuerySelectResult*>*)select
         json = [NSJSONSerialization dataWithJSONObject: root options: 0 error: &error];
         Assert(json, @"Failed to encode query as JSON: %@", error);
     }
-    return [self initWithDatabase: (CBLDatabase*)from.source JSONRepresentation: json];
+    
+    // TODO: cleanup?
+    if ([from.source isKindOfClass: [CBLDatabase class]])
+        return [self initWithDatabase: (CBLDatabase*)from.source JSONRepresentation: json];
+    else
+        return [self initWithCollection: (CBLCollection*)from.source JSONRepresentation: json];
```
[the remark this answers is not in the document]
Comment (jayahariv) [reply Objective-C/CBLQuery.mm:205]: I will add a TODO, if we use the defaultCollection now, it will start failing all the QueryTests.

**Comment by jayahariv** [comment]
Failure is not due to this PR, new commits on the lite-core-EE `master` branch causing failure with the current lite-core `staged` commit.

**Comment by pasin** [comment]
Isn't Lite-Core EE from iOS-EE's submodule?

**Comment by jayahariv** [comment]
No, its checking for branch-name(PR-##), or current target('master', release/XXX) 
https://github.com/couchbase/couchbase-lite-ios/blob/master/Jenkinsfile#L41

**Comment by pasin** [comment]
I don't remember why it clones LiteCore-EE directly instead of getting it from iOS-EE. It seems like it's better to get it from iOS-EE.

**Review by pasin — APPROVED** [review]

**Comment by jayahariv** [comment]
Since script was copied from LiteCore, which was using it to make sure, related `lite-core-EE` PR also taken into account. Which never failed till now, now realized that is not necessary, instead take it from `lite-ios-EE` submodule. 

Will make the change CBL-3208

**Pull request merged**